### ISSUE TITLE
Related issues of training

### ISSUE THREAD

**Issue by vhzy** [ZebangCheng/Emotion-LLaMA#31]
作者您好，感谢你们开源工作，我自己训练模型后发现无法复现您best_checkpoint的结果，因此有几个问题想要咨询您：
1.关于task_pool的选择，根据您的readme，stage1选择emotion和reason，stage2选择reason_v2，对应现在版本的代码，应该是stage1选择emotion和reason，stage2选择infer吗？
2.stage2应该使用stage1的哪个checkpoint？根据EmotionLLaMA/checkpoints/save_checkpoint/stage2/log.txt，您选择第七个checkpoint，readme写的是29个epoch，实际应该如何选择呢？
3.训练和测试的image_path是test3的视频数据即可吗？
4.论文中训练iter写的是训练300,000 steps，实际应该训练30*1000=30000个steps？这里是一个typo吗？
期待您的回复！

**Comment by ZebangCheng** [comment]
1.reason_v2相关部分已上传。感谢提醒。
2.readme中应该写的是，选择分数较高的checkpoint进行下一个阶段。为了方便，可以直接选用最后一个checkpoint。
3.“训练和测试的image_path是test3的视频数据即可吗？”这个问题不是很清楚？MER2023的训练集路径和测试集路径是一样的：
```
train:
    image_path: /home/user/big_space/datasets/Emotion/MER2023/video
    ann_path: /home/user/big_space/datasets/Emotion/MER2023/all_label_NCEV.txt
    
test:
    eval_file_path: /home/user/big_space/datasets/Emotion/MER2023/relative_test3_NCEV.txt # MER2023
    img_path: /home/user/big_space/datasets/Emotion/MER2023/video
```
4.300,000 steps 是估算的。300,000 steps = 30 epochs * 5000 iters_per_epoch * 2 (stages)

**Comment by vhzy** [comment]
> 1.reason_v2相关部分已上传。感谢提醒。 2.readme中应该写的是，选择分数较高的checkpoint进行下一个阶段。为了方便，可以直接选用最后一个checkpoint。 3.“训练和测试的image_path是test3的视频数据即可吗？”这个问题不是很清楚？MER2023的训练集路径和测试集路径是一样的：
> 
> ```
> train:
>     image_path: /home/user/big_space/datasets/Emotion/MER2023/video
>     ann_path: /home/user/big_space/datasets/Emotion/MER2023/all_label_NCEV.txt
>     
> test:
>     eval_file_path: /home/user/big_space/datasets/Emotion/MER2023/relative_test3_NCEV.txt # MER2023
>     img_path: /home/user/big_space/datasets/Emotion/MER2023/video
> ```
> 
> 4.300,000 steps 是估算的。300,000 steps = 30 epochs * 5000 iters_per_epoch * 2 (stages)

感谢您如此迅速的回复！
问题3:如果按照您的训练方式来，stage1的ann对应MERR_coarse_grained.txt,stage2的ann对应MERR_fine_grained.txt，image_path都使用MER2023 test3的video吗？（我理解的和实际看代码是这样）
问题4:好的呢我理解您的计算方式了，我看到您的train config之前是5000iter（注释掉了），现在是1000iter

**Comment by ZebangCheng** [comment]
3.是的，stage1的ann对应MERR_coarse_grained.txt,stage2的ann对应MERR_fine_grained.txt。因为MERR数据集的数据样本来自于MER2023 test3 semi无标签部分，所以都是同一个video目录。

**Comment by vhzy** [comment]
> 3.是的，stage1的ann对应MERR_coarse_grained.txt,stage2的ann对应MERR_fine_grained.txt。因为MERR数据集的数据样本来自于MER2023 test3 semi无标签部分，所以都是同一个video目录。

您好，我还想咨询一下stage2训练的学习率：
注意到您的log文件初始学习率是1e-5，但是您的yaml训练配置文件中，初始学习率，warmup学习率，最小学习率都是1e-6，应该将初始学习率设置成1e-5吗？

**Comment by vhzy** [comment]
作者您好，我训练模型仍然遇到一些问题，想向您请教，以下是MER2023f1 score的结果：
1.我使用您提供的demo模型，Emotion-Llama.pth（stage1的checkpoint），f1 score：0.81005
2.我使用您提供的best-checkpoint.pth（stage2的checkpoint），f1 score： 0.9032，基本符合论文结果
3.我自己训练stage1，f1 score:0.629，显然训练出现了一些问题
4.我使用您提供的stage1 Emotion-Llama.pth，进行第二阶段的instruction tuning，f1 score：0.8529，仍然无法复现论文结果
请问您有使用这一套代码重新训练模型，并可以复现论文结果吗？
以下是我的stage1的训练log：
{
    "run": {
        "task": "image_text_pretrain",
        "lr_sched": "linear_warmup_cosine_lr",
        "init_lr": 1e-05,
        "min_lr": 1e-06,
        "warmup_lr": 1e-06,
        "weight_decay": 0.05,
        "max_epoch": 30,
        "num_workers": 6,
        "iters_per_epoch": 1000,
        "warmup_steps": 1000,
        "seed": 42,
        "output_dir": "/data/hanzy/dataset/MER/Emotion-LLaMA/checkpoints/save_checkpoint/stage1",
        "amp": true,
        "resume_ckpt_path": null,
        "evaluate": false,
        "train_splits": [
            "train"
        ],
        "device": "cuda",
        "world_size": 4,
        "dist_url": "env://",
        "distributed": true,
        "wandb_log": false,
        "job_name": "minigptv2_finetune",
        "rank": 0,
        "gpu": 0,
        "dist_backend": "nccl"
    },
    "model": {
        "arch": "minigpt_v2",
        "image_size": 448,
        "drop_path_rate": 0,
        "use_grad_checkpoint": true,
        "vit_precision": "fp16",
        "freeze_vit": true,
        "prompt": "",
        "llama_model": "/data/hanzy/huggingface/Llama-2-7b-chat-hf",
        "lora_r": 64,
        "lora_alpha": 16,
        "model_type": "pretrain",
        "max_txt_len": 1024,
        "end_sym": "</s>",
        "ckpt": "/data/hanzy/dataset/MER/minigptv2_checkpoint.pth",
        "chat_template": true
    },
    "preprocess": {
        "vis_processor": {
            "train": {
                "name": "blip2_image_train",
                "image_size": 448
            },
            "eval": {
                "name": "blip2_image_eval",
                "image_size": 448
            }
        },
        "text_processor": {
            "train": {
                "name": "blip_caption"
            },
            "eval": {
                "name": "blip_caption"
            }
        }
    },
    "datasets": {
        "feature_face_caption": {
            "data_type": "images",
            "build_info": {
                "image_path": "/data/hanzy/dataset/MER/video/test3",
                "ann_path": "/data/hanzy/dataset/MER/MERR_coarse_grained.txt"
            },
            "batch_size": 1,
            "vis_processor": {
                "train": {
                    "name": "blip2_image_train",
                    "image_size": 448
                }
            },
            "text_processor": {
                "train": {
                    "name": "blip_caption"
                }
            },
            "sample_ratio": 30
        }
    }
}
{"train_lr": "0.000005", "train_loss": "0.489949"}
{"train_lr": "0.000010", "train_loss": "0.286261"}
{"train_lr": "0.000010", "train_loss": "0.245488"}
{"train_lr": "0.000010", "train_loss": "0.203850"}
{"train_lr": "0.000010", "train_loss": "0.185605"}
{"train_lr": "0.000009", "train_loss": "0.193672"}
{"train_lr": "0.000009", "train_loss": "0.176137"}
{"train_lr": "0.000009", "train_loss": "0.154514"}
{"train_lr": "0.000008", "train_loss": "0.139200"}
{"train_lr": "0.000008", "train_loss": "0.133122"}
{"train_lr": "0.000008", "train_loss": "0.127172"}
{"train_lr": "0.000007", "train_loss": "0.133997"}
{"train_lr": "0.000007", "train_loss": "0.128727"}
{"train_lr": "0.000006", "train_loss": "0.121930"}
{"train_lr": "0.000006", "train_loss": "0.090929"}
{"train_lr": "0.000005", "train_loss": "0.077853"}
{"train_lr": "0.000005", "train_loss": "0.078700"}
{"train_lr": "0.000004", "train_loss": "0.080075"}
{"train_lr": "0.000004", "train_loss": "0.078408"}
{"train_lr": "0.000003", "train_loss": "0.071077"}
{"train_lr": "0.000003", "train_loss": "0.068516"}
{"train_lr": "0.000003", "train_loss": "0.049539"}
{"train_lr": "0.000002", "train_loss": "0.033599"}
{"train_lr": "0.000002", "train_loss": "0.034416"}
{"train_lr": "0.000002", "train_loss": "0.036211"}
{"train_lr": "0.000001", "train_loss": "0.032698"}
{"train_lr": "0.000001", "train_loss": "0.031669"}
{"train_lr": "0.000001", "train_loss": "0.033036"}
{"train_lr": "0.000001", "train_loss": "0.024302"}
{"train_lr": "0.000001", "train_loss": "0.017018"}

**Comment by vhzy** [comment]
作者您好，我训练模型仍然遇到一些问题，想向您请教，以下是MER2023f1 score的结果：
1.我使用您提供的demo模型，Emotion-Llama.pth（stage1的checkpoint），f1 score：0.81005
2.我使用您提供的best-checkpoint.pth（stage2的checkpoint），f1 score： 0.9032，基本符合论文结果
3.我自己训练stage1，f1 score:0.629，显然训练出现了一些问题
4.我使用您提供的stage1 Emotion-Llama.pth，进行第二阶段的instruction tuning，f1 score：0.8529，仍然无法复现论文结果
请问您有使用这一套代码重新训练模型，并可以复现论文结果吗？
以下是我的stage1的训练log：
{
    "run": {
        "task": "image_text_pretrain",
        "lr_sched": "linear_warmup_cosine_lr",
        "init_lr": 1e-05,
        "min_lr": 1e-06,
        "warmup_lr": 1e-06,
        "weight_decay": 0.05,
        "max_epoch": 30,
        "num_workers": 6,
        "iters_per_epoch": 1000,
        "warmup_steps": 1000,
        "seed": 42,
        "output_dir": "/data/hanzy/dataset/MER/Emotion-LLaMA/checkpoints/save_checkpoint/stage1",
        "amp": true,
        "resume_ckpt_path": null,
        "evaluate": false,
        "train_splits": [
            "train"
        ],
        "device": "cuda",
        "world_size": 4,
        "dist_url": "env://",
        "distributed": true,
        "wandb_log": false,
        "job_name": "minigptv2_finetune",
        "rank": 0,
        "gpu": 0,
        "dist_backend": "nccl"
    },
    "model": {
        "arch": "minigpt_v2",
        "image_size": 448,
        "drop_path_rate": 0,
        "use_grad_checkpoint": true,
        "vit_precision": "fp16",
        "freeze_vit": true,
        "prompt": "",
        "llama_model": "/data/hanzy/huggingface/Llama-2-7b-chat-hf",
        "lora_r": 64,
        "lora_alpha": 16,
        "model_type": "pretrain",
        "max_txt_len": 1024,
        "end_sym": "</s>",
        "ckpt": "/data/hanzy/dataset/MER/minigptv2_checkpoint.pth",
        "chat_template": true
    },
    "preprocess": {
        "vis_processor": {
            "train": {
                "name": "blip2_image_train",
                "image_size": 448
            },
            "eval": {
                "name": "blip2_image_eval",
                "image_size": 448
            }
        },
        "text_processor": {
            "train": {
                "name": "blip_caption"
            },
            "eval": {
                "name": "blip_caption"
            }
        }
    },
    "datasets": {
        "feature_face_caption": {
            "data_type": "images",
            "build_info": {
                "image_path": "/data/hanzy/dataset/MER/video/test3",
                "ann_path": "/data/hanzy/dataset/MER/MERR_coarse_grained.txt"
            },
            "batch_size": 1,
            "vis_processor": {
                "train": {
                    "name": "blip2_image_train",
                    "image_size": 448
                }
            },
            "text_processor": {
                "train": {
                    "name": "blip_caption"
                }
            },
            "sample_ratio": 30
        }
    }
}
{"train_lr": "0.000005", "train_loss": "1.598095"}
{"train_lr": "0.000010", "train_loss": "0.639660"}
{"train_lr": "0.000010", "train_loss": "0.484230"}
{"train_lr": "0.000010", "train_loss": "0.412181"}
{"train_lr": "0.000010", "train_loss": "0.385341"}
{"train_lr": "0.000009", "train_loss": "0.380758"}
{"train_lr": "0.000009", "train_loss": "0.354872"}
{"train_lr": "0.000009", "train_loss": "0.343007"}
{"train_lr": "0.000008", "train_loss": "0.340674"}
{"train_lr": "0.000008", "train_loss": "0.325749"}
{"train_lr": "0.000008", "train_loss": "0.318956"}
{"train_lr": "0.000007", "train_loss": "0.318649"}
{"train_lr": "0.000007", "train_loss": "0.312097"}
{"train_lr": "0.000006", "train_loss": "0.306367"}
{"train_lr": "0.000006", "train_loss": "0.302833"}
{"train_lr": "0.000005", "train_loss": "0.290754"}
{"train_lr": "0.000005", "train_loss": "0.291068"}
{"train_lr": "0.000004", "train_loss": "0.287171"}
{"train_lr": "0.000004", "train_loss": "0.282526"}
{"train_lr": "0.000003", "train_loss": "0.272492"}
{"train_lr": "0.000003", "train_loss": "0.271695"}
{"train_lr": "0.000003", "train_loss": "0.271177"}
{"train_lr": "0.000002", "train_loss": "0.264974"}
{"train_lr": "0.000002", "train_loss": "0.262019"}
{"train_lr": "0.000002", "train_loss": "0.256479"}
{"train_lr": "0.000001", "train_loss": "0.253680"}
{"train_lr": "0.000001", "train_loss": "0.259529"}
{"train_lr": "0.000001", "train_loss": "0.258679"}
{"train_lr": "0.000001", "train_loss": "0.256160"}
{"train_lr": "0.000001", "train_loss": "0.254590"}

**Comment by vhzy** [comment]
作者您好，抱歉再次打扰，这几天在自己的机器上训练，仍然无法复现您best checkpoint的结果
1.我的计算环境：a800*4,cuda12.1，conda环境与您提供的yaml环境保持一致；请问您可以提供更多的计算环境信息，如cuda版本吗？
2.论文中的训练时长20h，是指单卡a100的卡时还是4卡a100的卡时呢？
3.我对您的工作非常感兴趣，对您花费精力维护repo表示非常respect，可以**微信**或者邮箱和您交流吗？

**Comment by ZebangCheng** [comment]
1. Emotion-LLaMA训练的环境是cuda11.7，也是在cuda11.7上测试分数。我们在升级到cuda12.6时，复现代码出现了轻微的分数下降（一两个样本预测结果和之前不一致），不影响实验结论。
2. 训练时4* 40G A100，每张卡batch_size为1. 
3. 可以的。先联系我邮箱吧。

> zebang.cheng@gmail.com

**Comment by vhzy** [comment]
感谢您的回复，我检查一下代码，之后用邮件和您沟通，再次向您表示感谢！

ZebangCheng ***@***.***> 于2024年12月6日周五 10:44写道：

**Comment by syc1336** [comment]
我沿用作者的设置，但是在42GB的卡上出现了爆显存的问题

![Image](https://github.com/user-attachments/assets/2850581d-7f38-403b-ad0a-e08393e7a17e)

**Comment by syc1336** [comment]
> > 1.reason_v2相关部分已上传。感谢提醒。 2.readme中应该写的是，选择分数较高的checkpoint进行下一个阶段。为了方便，可以直接选用最后一个checkpoint。 3.“训练和测试的image_path是test3的视频数据即可吗？”这个问题不是很清楚？MER2023的训练集路径和测试集路径是一样的：
> > ```
> > train:
> >     image_path: /home/user/big_space/datasets/Emotion/MER2023/video
> >     ann_path: /home/user/big_space/datasets/Emotion/MER2023/all_label_NCEV.txt
> >     
> > test:
> >     eval_file_path: /home/user/big_space/datasets/Emotion/MER2023/relative_test3_NCEV.txt # MER2023
> >     img_path: /home/user/big_space/datasets/Emotion/MER2023/video
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 4.300,000 steps 是估算的。300,000 steps = 30 epochs * 5000 iters_per_epoch * 2 (stages)
> 
> 感谢您如此迅速的回复！ 问题3:如果按照您的训练方式来，stage1的ann对应MERR_coarse_grained.txt,stage2的ann对应MERR_fine_grained.txt，image_path都使用MER2023 test3的video吗？（我理解的和实际看代码是这样） 问题4:好的呢我理解您的计算方式了，我看到您的train config之前是5000iter（注释掉了），现在是1000iter

我想问一下1000iter之后对模型训练有啥影响嘛？

**Comment by syc1336** [comment]
是否需要5000iter/epoch预训练才能复现作者的结果？

**Comment by syc1336** [comment]
我复现出的结果也存在了同样的问题

![Image](https://github.com/user-attachments/assets/98d7d708-295e-46af-a92b-f34406a29639)
